File: assistant-analyzer/src/main/java/ru/georgdeveloper/analyzer/EquipmentAnalyzer.java
package ru.georgdeveloper.analyzer;

import com.opencsv.CSVReader;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Component
public class EquipmentAnalyzer {
    
    private static final Logger logger = LoggerFactory.getLogger(EquipmentAnalyzer.class);
    private static final Pattern PROBLEM_PATTERN = Pattern.compile("Что Произошло:\\s*([^;]+)");
    private static final Pattern SOLUTION_PATTERN = Pattern.compile("Что ты сделал:\\s*([^;]+)");
    
    private final ObjectMapper objectMapper = new ObjectMapper();
    
    public List<RepairInstruction> analyzeEquipmentData(String csvFilePath) {
        logger.info("Начинаем анализ файла: {}", csvFilePath);
        
        Map<String, Map<String, Map<String, List<ProblemSolution>>>> groupedData = new HashMap<>();
        
        try (CSVReader reader = new CSVReader(new FileReader(csvFilePath))) {
            String[] headers = reader.readNext();
            Map<String, Integer> headerMap = createHeaderMap(headers);
            
            String[] row;
            while ((row = reader.readNext()) != null) {
                processRow(row, headerMap, groupedData);
            }
            
        } catch (Exception e) {
            logger.error("Ошибка чтения CSV файла: {}", e.getMessage());
            return Collections.emptyList();
        }
        
        return createRepairInstructions(groupedData);
    }
    
    private Map<String, Integer> createHeaderMap(String[] headers) {
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < headers.length; i++) {
            map.put(headers[i], i);
        }
        return map;
    }
    
    private void processRow(String[] row, Map<String, Integer> headerMap, 
                          Map<String, Map<String, Map<String, List<ProblemSolution>>>> groupedData) {
        
        String area = getValue(row, headerMap, "area", "Unknown");
        String machineName = getValue(row, headerMap, "machine_name", "Unknown");
        String mechanismNode = getValue(row, headerMap, "mechanism_node", "Unknown");
        String comments = getValue(row, headerMap, "comments", "");
        
        ProblemSolution ps = extractProblemAndSolution(comments);
        if (ps.isValid()) {
            groupedData.computeIfAbsent(area, k -> new HashMap<>())
                      .computeIfAbsent(machineName, k -> new HashMap<>())
                      .computeIfAbsent(mechanismNode, k -> new ArrayList<>())
                      .add(ps);
        }
    }
    
    private String getValue(String[] row, Map<String, Integer> headerMap, String column, String defaultValue) {
        Integer index = headerMap.get(column);
        return (index != null && index < row.length) ? row[index] : defaultValue;
    }
    
    private ProblemSolution extractProblemAndSolution(String comments) {
        if (comments == null || comments.trim().isEmpty()) {
            return new ProblemSolution("", "");
        }
        
        String problem = extractMatch(PROBLEM_PATTERN, comments);
        String solution = extractMatch(SOLUTION_PATTERN, comments);
        
        return new ProblemSolution(problem, solution);
    }
    
    private String extractMatch(Pattern pattern, String text) {
        Matcher matcher = pattern.matcher(text);
        return matcher.find() ? matcher.group(1).trim() : "";
    }
    
    private List<RepairInstruction> createRepairInstructions(
            Map<String, Map<String, Map<String, List<ProblemSolution>>>> groupedData) {
        
        List<RepairInstruction> instructions = new ArrayList<>();
        
        groupedData.forEach((area, machines) ->
            machines.forEach((machine, nodes) ->
                nodes.forEach((node, repairs) -> {
                    Map<String, Set<String>> problemGroups = repairs.stream()
                        .collect(Collectors.groupingBy(
                            ProblemSolution::getProblem,
                            Collectors.mapping(ProblemSolution::getSolution, Collectors.toSet())
                        ));
                    
                    problemGroups.forEach((problem, solutions) -> {
                        String combinedSolution = String.join("; ", solutions);
                        instructions.add(new RepairInstruction(area, machine, node, problem, combinedSolution));
                    });
                })
            )
        );
        
        logger.info("Создано {} инструкций по ремонту", instructions.size());
        return instructions;
    }
    
    public void saveToJson(List<RepairInstruction> instructions, String outputPath) {
        try {
            objectMapper.writerWithDefaultPrettyPrinter()
                       .writeValue(new File(outputPath), instructions);
            logger.info("Результат сохранен в файл: {}", outputPath);
        } catch (IOException e) {
            logger.error("Ошибка сохранения в JSON: {}", e.getMessage());
        }
    }
    
    private static class ProblemSolution {
        private final String problem;
        private final String solution;
        
        public ProblemSolution(String problem, String solution) {
            this.problem = problem;
            this.solution = solution;
        }
        
        public String getProblem() { return problem; }
        public String getSolution() { return solution; }
        public boolean isValid() { return !problem.isEmpty() && !solution.isEmpty(); }
    }
    
    public static class RepairInstruction {
        private String area;
        private String equipmentGroup;
        private String component;
        private String problem;
        private String solution;
        
        public RepairInstruction() {}
        
        public RepairInstruction(String area, String equipmentGroup, String component, 
                               String problem, String solution) {
            this.area = area;
            this.equipmentGroup = equipmentGroup;
            this.component = component;
            this.problem = problem;
            this.solution = solution;
        }
        
        // Getters and setters
        public String getArea() { return area; }
        public void setArea(String area) { this.area = area; }
        
        public String getEquipmentGroup() { return equipmentGroup; }
        public void setEquipmentGroup(String equipmentGroup) { this.equipmentGroup = equipmentGroup; }
        
        public String getComponent() { return component; }
        public void setComponent(String component) { this.component = component; }
        
        public String getProblem() { return problem; }
        public void setProblem(String problem) { this.problem = problem; }
        
        public String getSolution() { return solution; }
        public void setSolution(String solution) { this.solution = solution; }
    }
}

File: assistant-analyzer/equipment_analyzer.py
#!/usr/bin/env python3
"""
Оптимизированный анализатор оборудования
"""

import csv
import json
import re
from collections import defaultdict
from pathlib import Path
from typing import Dict, List, Tuple, Optional

class EquipmentAnalyzer:
    """Анализатор данных оборудования"""
    
    PROBLEM_PATTERN = re.compile(r'Что Произошло:\s*([^;]+)')
    SOLUTION_PATTERN = re.compile(r'Что ты сделал:\s*([^;]+)')
    
    def __init__(self, csv_file_path: str):
        self.csv_file_path = Path(csv_file_path)
        self.repair_instructions = []
    
    def extract_problem_solution(self, comments: str) -> Tuple[str, str]:
        """Извлекает проблему и решение из комментариев"""
        if not comments:
            return "", ""
        
        problem_match = self.PROBLEM_PATTERN.search(comments)
        solution_match = self.SOLUTION_PATTERN.search(comments)
        
        problem = problem_match.group(1).strip() if problem_match else ""
        solution = solution_match.group(1).strip() if solution_match else ""
        
        return problem, solution
    
    def analyze(self) -> List[Dict]:
        """Анализирует CSV файл и создает инструкции по ремонту"""
        grouped_data = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
        
        try:
            with open(self.csv_file_path, 'r', encoding='utf-8') as csvfile:
                reader = csv.DictReader(csvfile)
                
                for row in reader:
                    area = row.get('area', 'Unknown')
                    machine_name = row.get('machine_name', 'Unknown')
                    mechanism_node = row.get('mechanism_node', 'Unknown')
                    comments = row.get('comments', '')
                    
                    problem, solution = self.extract_problem_solution(comments)
                    
                    if problem and solution:
                        grouped_data[area][machine_name][mechanism_node].append({
                            'problem': problem,
                            'solution': solution
                        })
        
        except FileNotFoundError:
            raise FileNotFoundError(f"Файл не найден: {self.csv_file_path}")
        except Exception as e:
            raise Exception(f"Ошибка чтения файла: {e}")
        
        return self._create_instructions(grouped_data)
    
    def _create_instructions(self, grouped_data) -> List[Dict]:
        """Создает инструкции по ремонту из сгруппированных данных"""
        instructions = []
        
        for area, machines in grouped_data.items():
            for machine, nodes in machines.items():
                for node, repairs in nodes.items():
                    problem_groups = defaultdict(set)
                    
                    for repair in repairs:
                        problem_groups[repair['problem']].add(repair['solution'])
                    
                    for problem, solutions in problem_groups.items():
                        instructions.append({
                            "area": area,
                            "equipment_group": machine,
                            "component": node,
                            "problem": problem,
                            "solution": "; ".join(solutions)
                        })
        
        return instructions
    
    def save_to_json(self, instructions: List[Dict], output_path: str) -> None:
        """Сохраняет инструкции в JSON файл"""
        with open(output_path, 'w', encoding='utf-8') as f:
            json.dump(instructions, f, ensure_ascii=False, indent=2)

def main():
    """Основная функция для запуска анализа"""
    csv_file = "assistant-core/src/main/resources/equipment_maintenance_records_202508150901.csv"
    output_file = "repair_instructions.json"
    
    try:
        analyzer = EquipmentAnalyzer(csv_file)
        instructions = analyzer.analyze()
        analyzer.save_to_json(instructions, output_file)
        
        print(f"Анализ завершен. Создано {len(instructions)} инструкций.")
        print(f"Результат сохранен в: {output_file}")
        
    except Exception as e:
        print(f"Ошибка: {e}")

if __name__ == "__main__":
    main()